src/DotNetty.Handlers/Tls/TlsHandler.cs

```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
 
         protected override void Decode(IChannelHandlerContext context, IByteBuffer input, List<object> output)
         {
-            // pass bytes to SslStream through input -> trigger HandleSslStreamRead. After this call sslStreamReadBuffer may or may not have bytes to read
-            this.mediationStream.AcceptBytes(input);
+            int startOffset = input.ReaderIndex;
+            int endOffset = input.WriterIndex;
+            int offset = startOffset;
+            int totalLength = 0;
 
-            if (!this.EnsureAuthenticated())
+            List<int> packetLengths;
+            // if we calculated the length of the current SSL record before, use that information.
+            if (this.packetLength > 0)
             {
-                return;
+                if (endOffset - startOffset < this.packetLength)
+                {
+                    return;
+                }
+                else
+                {
+                    packetLengths = new List<int>(4);
+                    packetLengths.Add(this.packetLength);
+                    offset += this.packetLength;
+                    totalLength = this.packetLength;
+                    this.packetLength = 0;
+                }
+            }
+            else
+            {
+                packetLengths = new List<int>(4);
```

Comment:
btw, I believe this is the default length of an initialized List.

Also, can we recycle / reuse this across calls?